File: services/cookbook/core/src/main/java/family/haschka/wolkenschloss/cookbook/recipe/WrappedString.java
package family.haschka.wolkenschloss.cookbook.recipe;

import java.util.Arrays;
import java.util.stream.Collectors;

public record WrappedString(String text) {

    public String wrap(int i) {
        var paragraphs = text.split("\n\n");

        return Arrays.stream(paragraphs)
                .map(p -> wrapParagraph(p, i))
                .collect(Collectors.joining("\n\n"));
    }

    private String wrapParagraph(String paragraph, int i) {
        var x = paragraph.split("\\s+");

        StringBuilder builder = new StringBuilder();

        int row = 0;
        int pos = 0;
        for (int idx = 0; idx < x.length; idx++) {

            // Wort hinzufügen
            builder.append(x[idx]);
            row = row + x[idx].length();
            pos = pos + x[idx].length();

            // Leerzeichen hinzufügen, wenn noch weitere Wörter folgen,
            // die nicht zu einem Zeilenumbruch führen.
            if (idx + 1 < x.length) {
                if (row + 1 + x[idx + 1].length()  > i) {
                    builder.append("\n");
                    row = 0;
                } else {
                    builder.append(" ");
                    row = row + 1;
                    pos = pos + 1;
                }
            }
        }

        return builder.toString();
    }
}

File: services/cookbook/core/src/main/kotlin/recipe/WrappedString.kt
package family.haschka.wolkenschloss.cookbook.recipe

data class WrappedString(val text: String) {

    fun wrap(i: Int): String {
        val paragraphs = text.split("\n\n")
        return paragraphs.joinToString("\n\n") { p -> wrapParagraph(p, i) }
    }

    private fun wrapParagraph(paragraph: String, i: Int): String {
        val regex = "\\s+".toRegex()
        val x = paragraph.split(regex)

        val builder = StringBuilder()

        var row = 0
        var pos = 0
        for (idx: Int in x.indices) {

            // Wort hinzufügen
            builder.append(x[idx])
            row += x[idx].length
            pos += x[idx].length

            // Leerzeichen hinzufügen, wenn noch weitere Wörter folgen,
            // die nicht zu einem Zeilenumbruch führen.
            if (idx + 1 < x.size) {
                if (row + 1 + x[idx + 1].length > i) {
                    builder.append("\n")
                    row = 0
                } else {
                    builder.append(" ")
                    row += 1
                    pos += 1
                }
            }
        }

        return builder.toString()
    }
}
